service/matching/taskWriter.go

```diff
@@ -55,56 +56,74 @@ type (
 
 	// taskWriter writes tasks sequentially to persistence
 	taskWriter struct {
+		status       int32
 		tlMgr        *taskQueueManagerImpl
 		config       *taskQueueConfig
 		taskQueueID  *taskQueueID
 		appendCh     chan *writeTaskRequest
 		taskIDBlock  taskIDBlock
 		maxReadLevel int64
-		stopped      int64 // set to 1 if the writer is stopped or is shutting down
 		logger       log.Logger
-		stopCh       chan struct{} // shutdown signal for all routines in this class
+		shutdownChan chan struct{}
 	}
 )
 
 // errShutdown indicates that the task queue is shutting down
 var errShutdown = &persistence.ConditionFailedError{Msg: "task queue shutting down"}
 
-func newTaskWriter(tlMgr *taskQueueManagerImpl) *taskWriter {
+func newTaskWriter(
+	tlMgr *taskQueueManagerImpl,
+	block taskIDBlock,
+) *taskWriter {
 	return &taskWriter{
-		tlMgr:       tlMgr,
-		config:      tlMgr.config,
-		taskQueueID: tlMgr.taskQueueID,
-		stopCh:      make(chan struct{}),
-		appendCh:    make(chan *writeTaskRequest, tlMgr.config.OutstandingTaskAppendsThreshold()),
-		logger:      tlMgr.logger,
+		status:       common.DaemonStatusInitialized,
+		tlMgr:        tlMgr,
+		config:       tlMgr.config,
+		taskQueueID:  tlMgr.taskQueueID,
+		appendCh:     make(chan *writeTaskRequest, tlMgr.config.OutstandingTaskAppendsThreshold()),
+		taskIDBlock:  block,
+		maxReadLevel: block.start - 1,
+		logger:       tlMgr.logger,
+
+		shutdownChan: make(chan struct{}),
 	}
 }
 
-func (w *taskWriter) Start(block taskIDBlock) {
-	w.taskIDBlock = block
-	w.maxReadLevel = block.start - 1
+func (w *taskWriter) Start() {
+	if !atomic.CompareAndSwapInt32(
+		&w.status,
+		common.DaemonStatusInitialized,
+		common.DaemonStatusStarted,
+	) {
+		return
+	}
+
 	go w.taskWriterLoop()
 }
 
 // Stop stops the taskWriter
 func (w *taskWriter) Stop() {
-	if atomic.CompareAndSwapInt64(&w.stopped, 0, 1) {
-		close(w.stopCh)
+	if !atomic.CompareAndSwapInt32(
+		&w.status,
+		common.DaemonStatusStarted,
```

Comment:
Just checking if initialized is a valid state to stop?

Reply:
only stop if state is `started`
only start if state is `initialized`
in case `Start` or `Stop` called multiple times